classes/Hook.php

```diff
@@ -642,9 +642,28 @@ public static function getHookModuleExecList($hook_name = null)
                     $sql->where('((h.`name` = "displayPayment" OR h.`name` = "displayPaymentEU" OR h.`name` = "paymentOptions") AND (SELECT `id_currency` FROM `' . _DB_PREFIX_ . 'module_currency` mcr WHERE mcr.`id_module` = m.`id_module` AND `id_currency` IN (' . (int) $context->currency->id . ', -1, -2) LIMIT 1) IN (' . (int) $context->currency->id . ', -1, -2))');
                 }
                 if (Validate::isLoadedObject($context->cart)) {
-                    $carrier = new Carrier($context->cart->id_carrier);
-                    if (Validate::isLoadedObject($carrier)) {
-                        $sql->where('((h.`name` = "displayPayment" OR h.`name` = "displayPaymentEU" OR h.`name` = "paymentOptions") AND (SELECT `id_reference` FROM `' . _DB_PREFIX_ . 'module_carrier` mcar WHERE mcar.`id_module` = m.`id_module` AND `id_reference` = ' . (int) $carrier->id_reference . ' AND `id_shop` = ' . (int) $context->shop->id . ' LIMIT 1) = ' . (int) $carrier->id_reference . ')');
+                    $SQLsubquery = '';
+                    $carriers = [];
+                    $cartCarriers = $context->cart->getDeliveryOption();
```

Comment:
Maybe it would be nice to put all of this logic into a dedicated function, what do you think ? Something like :

```
if (Validate::isLoadedObject($context->cart)) {
    $this->addCarrierSqlCondition($sql);
)
...

protected function addCarrierSqlCondition(DbQuery $sqlQuery) {
    $subquery = '';
    $carriers = [];
    $cartCarriers = $context->cart->getDeliveryOption();
    ...
}
```

Reply:
Yes, it could be a better solution, isolating on that way that code and keeping clean the main function. If you check the fix as a valid I do not mind to create that function.